rust/lance/src/dataset.rs

```diff
@@ -822,70 +822,134 @@ impl Dataset {
-            let local_row_ids = row_ids_per_fragment.get(&(f.id() as u64))?;
-            Some((f, local_row_ids))
-        });
-
-        let projection_with_row_id = Schema::merge(
-            projection,
-            &ArrowSchema::new(vec![ArrowField::new(
-                ROW_ID,
-                arrow::datatypes::DataType::UInt64,
-                false,
-            )]),
-        )?;
-        let schema_with_row_id = Arc::new(ArrowSchema::from(&projection_with_row_id));
+        if row_ids.is_empty() {
+            return Ok(RecordBatch::new_empty(Arc::new(projection.into())));
+        }
 
-        let batches = stream::iter(fragment_and_indices)
-            .then(|(fragment, indices)| fragment.take_rows(indices, projection, true))
-            .try_collect::<Vec<_>>()
-            .await?;
+        let row_id_meta = check_row_ids(row_ids);
 
-        let one_batch = concat_batches(&schema_with_row_id, &batches)?;
-        // Note: one_batch may contains fewer rows than the number of requested
-        // row ids because some rows may have been deleted. Because of this, we
-        // get the results with row ids so that we can re-order the results
-        // to match the requested order.
-
-        let returned_row_ids = one_batch
-            .column_by_name(ROW_ID)
-            .ok_or_else(|| Error::Internal {
-                message: "ROW_ID column not found".into(),
-            })?
-            .as_any()
-            .downcast_ref::<UInt64Array>()
-            .ok_or_else(|| Error::Internal {
-                message: "ROW_ID column is not UInt64Array".into(),
-            })?
-            .values();
-
-        let remapping_index: UInt64Array = row_ids
-            .iter()
-            .filter_map(|o| returned_row_ids.binary_search(o).ok().map(|pos| pos as u64))
-            .collect();
+        if row_id_meta.contiguous {
+            // Fastest path: Can use `read_range` directly
+            let start = row_ids.first().expect("empty range passed to take_rows");
+            let fragment_id = (start >> 32) as usize;
+            let range_start = *start as u32 as usize;
+            let range_end =
+                *row_ids.last().expect("empty range passed to take_rows") as u32 as usize;
+            let range = range_start..(range_end + 1);
 
-        // Remove the row id column.
-        let keep_indices = (0..one_batch.num_columns() - 1).collect::<Vec<_>>();
-        let one_batch = one_batch.project(&keep_indices)?;
-        let struct_arr: StructArray = one_batch.into();
-        let reordered = take(&struct_arr, &remapping_index, None)?;
-        Ok(as_struct_array(&reordered).into())
+            let fragment = self.get_fragment(fragment_id).ok_or_else(|| {
+                Error::invalid_input(format!("row_id belongs to non-existant fragment: {start}"))
+            })?;
+
+            let reader = fragment.open(projection).await?;
+            reader.read_range(range).await
```

Comment:
It does not. But in order for a contiguous range to span multiple fragments is if one fragment has `u32::MAX` rows (4.2 billion). That might be an edge case we should handle.

Reply:
Does a contiguous range have to start at 0?  Couldn't a range of 2 rows be "the last row from fragment 5 and the first row in fragment 6"?

Reply:
Oh, I see.  There will be a gap in the row ids because the last row from fragment 5 will be `5 * 2^32 + frag_length` and the first row in fragment 6 will be `6* 2^32`.